### ISSUE TITLE
Support psql meta commands

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#4265]
`psql` supports a long list of [meta-commands](https://www.postgresql.org/docs/10/app-psql.html). Eventually, we should support all of them. In the short term, we should support the most popular ones!

This is intended to be a tracking issue for adding support for individual commands.

- [x] `\l`: https://github.com/MaterializeInc/materialize/issues/2156
- [x] `\d`: https://github.com/MaterializeInc/materialize/issues/4279
- [x] `\dv`: https://github.com/MaterializeInc/materialize/issues/4283
- [x] `\dt`: https://github.com/MaterializeInc/materialize/issues/4284
- [x] `\di`: https://github.com/MaterializeInc/materialize/issues/4285

Out of scope for this issue:
- `\copy`: https://github.com/MaterializeInc/materialize/issues/3324

**Comment by mjibson** [comment]
Things I use most commonly are `show create` `show databases` and `show tables/views`. I think that maps to `\l` and `\d`. `show create` doesn't have a meta command afaik.

**Comment by rjnn** [comment]
The comprehensive list can be obtained by firing up psql, typing `\` and tab completing:

```         
arjun=# \
\!             \dFt           \dg            \endif         \p
\?             \dL            \di            \errverbose    \password
\C             \dO            \dl            \ev            \prompt
\H             \dP            \dm            \f             \pset
\T             \dPi           \dn            \g             \q
\a             \dPt           \do            \gdesc         \qecho
\cd            \dRp           \dp            \gexec         \r
\connect       \dRs           \drds          \gset          \s
\conninfo      \dS            \ds            \gx            \set
\copy          \dT            \dt            \h             \setenv
\copyright     \da            \du            \help          \sf
\crosstabview  \db            \dv            \i             \sv
\d             \dc            \dx            \if            \t
\dA            \dd            \dy            \ir            \timing
\dC            \ddp           \e             \l             \unset
\dD            \des           \echo          \lo_export     \w
\dE            \det           \ef            \lo_import     \watch
\dF            \deu           \elif          \lo_list       \x
\dFd           \dew           \else          \lo_unlink     \z
\dFp           \df            \encoding      \o             
```

When psql is started with the -E or --echo-hidden options, it will display the query associated with a backslash command before executing that command.

That can be used to produce a good burndown list of all the `pg_catalog` queries

**Comment by benesch** [comment]
The output of `\?` is much more useful, IMO:

```
General
  \copyright             show PostgreSQL usage and distribution terms
  \crosstabview [COLUMNS] execute query and display results in crosstab
  \errverbose            show most recent error message at maximum verbosity
  \g [FILE] or ;         execute query (and send results to file or |pipe)
  \gdesc                 describe result of query, without executing it
  \gexec                 execute query, then execute each value in its result
  \gset [PREFIX]         execute query and store results in psql variables
  \gx [FILE]             as \g, but forces expanded output mode
  \q                     quit psql
  \watch [SEC]           execute query every SEC seconds

Help
  \? [commands]          show help on backslash commands
  \? options             show help on psql command-line options
  \? variables           show help on special variables
  \h [NAME]              help on syntax of SQL commands, * for all commands

Query Buffer
  \e [FILE] [LINE]       edit the query buffer (or file) with external editor
  \ef [FUNCNAME [LINE]]  edit function definition with external editor
  \ev [VIEWNAME [LINE]]  edit view definition with external editor
  \p                     show the contents of the query buffer
  \r                     reset (clear) the query buffer
  \s [FILE]              display history or save it to file
  \w FILE                write query buffer to file

Input/Output
  \copy ...              perform SQL COPY with data stream to the client host
  \echo [STRING]         write string to standard output
  \i FILE                execute commands from file
  \ir FILE               as \i, but relative to location of current script
  \o [FILE]              send all query results to file or |pipe
  \qecho [STRING]        write string to query output stream (see \o)

Conditional
  \if EXPR               begin conditional block
  \elif EXPR             alternative within current conditional block
  \else                  final alternative within current conditional block
  \endif                 end conditional block

Informational
  (options: S = show system objects, + = additional detail)
  \d[S+]                 list tables, views, and sequences
  \d[S+]  NAME           describe table, view, sequence, or index
  \da[S]  [PATTERN]      list aggregates
  \dA[+]  [PATTERN]      list access methods
  \db[+]  [PATTERN]      list tablespaces
  \dc[S+] [PATTERN]      list conversions
  \dC[+]  [PATTERN]      list casts
  \dd[S]  [PATTERN]      show object descriptions not displayed elsewhere
  \dD[S+] [PATTERN]      list domains
  \ddp    [PATTERN]      list default privileges
  \dE[S+] [PATTERN]      list foreign tables
  \det[+] [PATTERN]      list foreign tables
  \des[+] [PATTERN]      list foreign servers
  \deu[+] [PATTERN]      list user mappings
  \dew[+] [PATTERN]      list foreign-data wrappers
  \df[anptw][S+] [PATRN] list [only agg/normal/procedures/trigger/window] functions
  \dF[+]  [PATTERN]      list text search configurations
  \dFd[+] [PATTERN]      list text search dictionaries
  \dFp[+] [PATTERN]      list text search parsers
  \dFt[+] [PATTERN]      list text search templates
  \dg[S+] [PATTERN]      list roles
  \di[S+] [PATTERN]      list indexes
  \dl                    list large objects, same as \lo_list
  \dL[S+] [PATTERN]      list procedural languages
  \dm[S+] [PATTERN]      list materialized views
  \dn[S+] [PATTERN]      list schemas
  \do[S]  [PATTERN]      list operators
  \dO[S+] [PATTERN]      list collations
  \dp     [PATTERN]      list table, view, and sequence access privileges
  \dP[itn+] [PATTERN]    list [only index/table] partitioned relations [n=nested]
  \drds [PATRN1 [PATRN2]] list per-database role settings
  \dRp[+] [PATTERN]      list replication publications
  \dRs[+] [PATTERN]      list replication subscriptions
  \ds[S+] [PATTERN]      list sequences
  \dt[S+] [PATTERN]      list tables
  \dT[S+] [PATTERN]      list data types
  \du[S+] [PATTERN]      list roles
  \dv[S+] [PATTERN]      list views
  \dx[+]  [PATTERN]      list extensions
  \dy     [PATTERN]      list event triggers
  \l[+]   [PATTERN]      list databases
  \sf[+]  FUNCNAME       show a function's definition
  \sv[+]  VIEWNAME       show a view's definition
  \z      [PATTERN]      same as \dp

Formatting
  \a                     toggle between unaligned and aligned output mode
  \C [STRING]            set table title, or unset if none
  \f [STRING]            show or set field separator for unaligned query output
  \H                     toggle HTML output mode (currently off)
  \pset [NAME [VALUE]]   set table output option
                         (border|columns|csv_fieldsep|expanded|fieldsep|
                         fieldsep_zero|footer|format|linestyle|null|
                         numericlocale|pager|pager_min_lines|recordsep|
                         recordsep_zero|tableattr|title|tuples_only|
                         unicode_border_linestyle|unicode_column_linestyle|
                         unicode_header_linestyle)
  \t [on|off]            show only rows (currently off)
  \T [STRING]            set HTML <table> tag attributes, or unset if none
  \x [on|off|auto]       toggle expanded output (currently off)

Connection
  \c[onnect] {[DBNAME|- USER|- HOST|- PORT|-] | conninfo}
                         connect to new database (currently "benesch")
  \conninfo              display information about current connection
  \encoding [ENCODING]   show or set client encoding
  \password [USERNAME]   securely change the password for a user

Operating System
  \cd [DIR]              change the current working directory
  \setenv NAME [VALUE]   set or unset environment variable
  \timing [on|off]       toggle timing of commands (currently off)
  \! [COMMAND]           execute command in shell or start interactive shell

Variables
  \prompt [TEXT] NAME    prompt user to set internal variable
  \set [NAME [VALUE]]    set internal variable, or list all if no parameters
  \unset NAME            unset (delete) internal variable

Large Objects
  \lo_export LOBOID FILE
  \lo_import FILE [COMMENT]
  \lo_list
  \lo_unlink LOBOID      large object operations
```

The only relevant commands to us are the "informational" commands, as everything else is handled client side by psql. That's everything starting with `\d...` plus `\l`, `\sf`, `\sv`, and `\z`.

I'd like to propose that we consider this issue closed when the following commands are supported:

- [ ] `\l`
- [ ] `\d`
- [ ] `\dv`
- [ ] `\dt`
- [ ] `\di`

Otherwise the list is so long that this tracking issue is going to stay open forever.

**Comment by JLDLaughlin** [comment]
Currently blocked by: https://github.com/MaterializeInc/materialize/issues/4289

**Comment by JLDLaughlin** [comment]
All of the in-scope meta commands are complete!
